Data/HashMap/Base.hs

```diff
@@ -1822,6 +1800,58 @@ updateOrConcatWithKey f ary1 ary2 = A.run $ do
     return mary
 {-# INLINABLE updateOrConcatWithKey #-}
 
+-- Helper functions for very rare cases.
+
+-- Remove (any) one element from an array of hashmaps
+unConsA :: A.Array (HashMap k v) -> UnCons k v
+unConsA ary = case A.length ary of
+  1 -> case A.index ary 0 of
+    Empty -> error "Invariant violated: Empty hashmap in array."
+    Leaf _ l -> NowEmpty l
+    BitmapIndexed bm ary' -> case unConsA ary' of
+      NowEmpty l -> NowEmpty l
+      UnConsed _ l a' -> UnConsed 0 l (A.singleton (BitmapIndexed bm a'))
+    Full ary' -> case unConsA ary' of
+        NowEmpty l -> NowEmpty l
+        UnConsed i l a' ->
+          let bm = fullNodeMask .&. complement (1 `unsafeShiftL` i)
+          in UnConsed 0 l (A.singleton (BitmapIndexed bm a'))
+    Collision h l1 l2 hm -> UnConsed 0 l1 $ A.singleton $ case unConsHM hm of
+      Nothing -> Leaf h l2
+      Just (l3, hm') -> Collision h l2 l3 hm'
+  _ -> goA 0
+  where
+    goA ix = case unConsHM $ A.index ary ix of
+      Nothing -> goA $ ix + 1 -- This is fine because the array only houses nonempty hashmaps.
+      Just (l, hm') -> case hm' of
+        Empty -> UnConsed ix l $ A.delete ary ix -- This is fine because the array cannot become empty.
+        _ -> UnConsed ix l $ A.update ary ix hm'
+
+data UnCons k v
+  = NowEmpty (Leaf k v)
+  | UnConsed
+    Int -- The indexed of the array that was un-consed from
+    (Leaf k v) -- The leaf that was un-consed
+    (A.Array (HashMap k v)) -- The leftover array
+  deriving (Show)
+
+-- Remove (any) one element from a hashmap
+unConsHM :: HashMap k v -> Maybe (Leaf k v, HashMap k v)
+unConsHM hm = case hm of
+  Empty -> Nothing
+  Leaf _ l -> Just (l, Empty)
+  BitmapIndexed bm ary' -> case unConsA ary' of
+    NowEmpty l -> Just (l, Empty)
+    UnConsed _ l a' -> Just (l, BitmapIndexed bm a') -- TODO ignore the index?
```

Comment:
We need to distinguish between the `UnconsedFromFirstA` and `RemovedFirstA` possibilities. If we `RemovedFirstA`, then we need to update the bitmap.